File: Cs50/General Projects/pset6/sentimental-credit/credit.py
# RROTUNDU 2023

import sys


# Main function
def main():
    # Get user input as string for determining length
    numberString = input("Number: ")

    # Variable will hold the card type
    CardType = ""

    # Verify the card type
    if(isAmex(numberString)):
        CardType = "AMEX"

        if(not checkSum(numberString)):
            CardType = "INVALID"

    elif(isMaster(numberString)):
        CardType = "MASTERCARD"

        if(not checkSum(numberString)):
            CardType = "INVALID"

    elif(isVisa(numberString)):
        CardType = "VISA"

        if(not checkSum(numberString)):
            CardType = "INVALID"
    else:
        CardType = "INVALID"

    print("Number: " + numberString + "\n" + CardType)







# Boolean is it AMEX
def isAmex(numstr):
    # The length of the given string
    length = len(numstr)
    # That string converted to a number(should be all numbers anyways)
    number = int(numstr)

    if(length == 15):
        # Pow10 variables
        powTen = pow(10, length)
        divisor = pow(10, (length - 2))

        # Modulo of the last 2 digits
        lastTwo  = (number % powTen) / divisor

        if(lastTwo == 34 or lastTwo == 37):
            return True
        else:
            return False
    else:
        return False






# Boolean is it Mastercard
def isMaster(numstr):
    # The length of the given string
    length = len(numstr)
    # That string converted to a number(should be all numbers anyways)
    number = int(numstr)

    # Check length first
    if(length == 16):
         # Pow10 variables
        powTen = pow(10, length)
        divisor = pow(10, (length - 2))

        # Modulo of the last 2 digits
        lastTwo  = (number % powTen) / divisor
        #if (LastTwo == 51 || LastTwo == 52 || LastTwo == 53 || LastTwo == 54 || LastTwo == 55)
        if(lastTwo == 51 or lastTwo == 52 or lastTwo == 53 or lastTwo == 54):
            return True
        else:
            return False
    else:
        return False





# Boolean is it VISA
def isVisa(numstr):
    # The length of the given string
    length = len(numstr)
    # That string converted to a number(should be all numbers anyways)
    number = int(numstr)

    if(length == 13 or length == 16):
     # Pow10 variables
        powTen = pow(10, length)
        divisor = pow(10, (length - 1))

        lastOne = ((number % powTen) / divisor)

        if(lastOne == 4):
            return True
        else:
            return False
    else:
        return False







# Boolean is the checksum correct
def checkSum(numstr):
    length = len(numstr)
    longNum = int(numstr)
    total = otherSum(longNum, length) + lastSum(longNum, length)
    if (total % 10 == 0):
        return true
    else:
        return false



# Int does the other sum
def otherSum(inNum, length):
    x = 2
    sum = 0
    for i in len(length):
        pwr = pow(10, x)
        div = pow(10, x - 1)
        current = ((inNum % pwr) / div)
        #digit is multiplied by two
        current = current * 2
        #the result is then broken into individual digits, and added together

        z = 1
        while (z < 3):
            subpwr = pow(10, z)
            subdiv = pow(10, z - 1)
            subcurrent = ((current % subpwr) / subdiv)
            sum = sum + subcurrent
            z = z + 1
        x = x + 2
    return sum;

# Int does the final sum
def lastSum(inNum, length):
    x = 1
    sum = 0

    for i in length:
        pwr = pow(10, x)
        div = pow(10, x - 1)
        current = ((inNum % pwr) / div)
        #result is summed together
        sum = sum + current
        # printf("CURRENT: %li\n", current);
        #printf("SUM: %i\n", sum);
        x = x + 2
    return sum

main()

File: Cs50/General Projects/pset1/Credit/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdbool.h>

//Raz Rotundu 2022

//Verify first two and number length
bool IsAmex(string numstr);
bool IsMaster(string numstr);
bool IsVisa(string numstr);

//do the checksum operation
bool checksum(string numstr);
//checksum functions
int otherSum(long in, int len);
int lastSum(long in, int len);

int main(void)
{
    //get user input as string(for determining length)
    string numberString = get_string("Number: ");
    //card type
    string CardType;
    //verify card type
    //bool validity = IsAmex(numberString);
    if (IsAmex(numberString) == true)
    {
        CardType = "AMEX";
        if (checksum(numberString) != true)
        {
            CardType = "INVALID";
        }
    }
    else if (IsMaster(numberString) == true)
    {
        CardType = "MASTERCARD";
        if (checksum(numberString) != true)
        {
            CardType = "INVALID";
        }
    }
    else if (IsVisa(numberString) == true)
    {
        CardType = "VISA";
        if (checksum(numberString) != true)
        {
            CardType = "INVALID";
        }
    }
    else
    {
        CardType = "INVALID";
    }
    printf("Number: %s\n", CardType);
}

//AMEX definition
bool IsAmex(string numstr)
{
    int length = strlen(numstr);
    long number = atol(numstr);
    //check the lenght first
    if (length == 15)
    {
        //checks the last two digits next
        long PowTen = pow(10, length);
        long divisor = pow(10, (length - 2));
        //modulo for the last two digitd
        long LastTwo = ((number % PowTen) / divisor);
        if (LastTwo == 34 || LastTwo == 37)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
    else
    {
        return false;
    }
    //return false by default
    return false;
}
//MASTER definition
bool IsMaster(string numstr)
{
    int length = strlen(numstr);
    long number = atol(numstr);
    //check the lenght first
    if (length == 16)
    {
        //checks the last two digits next
        long PowTen = pow(10, length);
        long divisor = pow(10, (length - 2));
        //modulo for the last two digitd
        long LastTwo = ((number % PowTen) / divisor);
        if (LastTwo == 51 || LastTwo == 52 || LastTwo == 53 || LastTwo == 54 || LastTwo == 55)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
    else
    {
        return false;
    }
    //return false by default
    return false;
}
//VISA Definition
bool IsVisa(string numstr)
{
    int length = strlen(numstr);
    long number = atol(numstr);
    //check the lenght first
    if (length == 13 || length == 16)
    {
        //checks the last two digits next
        long PowTen = pow(10, length);
        long divisor = pow(10, (length - 1));
        //modulo for the last two digitd
        long LastOne = ((number % PowTen) / divisor);
        if (LastOne == 4)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
    else
    {
        return false;
    }
    //return false by default
    return false;
}
bool checksum(string numstr)
{
    //length of num
    int length = strlen(numstr);
    //convert to long
    long longNum = atol(numstr);
    int total = otherSum(longNum, length) + lastSum(longNum, length);
    if (total % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

int otherSum(long in, int len)
{
    //exponent for finding every other number
    int x = 2;
    int sum = 0;
    //selects every other integer, starting from second last
    for (int i = 0; i < len; i++)
    {
        long pwr = pow(10, x);
        long div = pow(10, x - 1);
        long current = ((in % pwr) / div);
        //digit is multiplied by two
        current = current * 2;
        //the result is then broken into individual digits, and added together
        for (int z = 1; z < 3; z++)
        {
            long subpwr = pow(10, z);
            long subdiv = pow(10, z - 1);
            long subcurrent = ((current % subpwr) / subdiv);
            sum = sum + subcurrent;
            //printf("Adding: %li\n", subcurrent);

        }
        //printf("SUM: %i\n", sum);
        x = x + 2;
    }
    return sum;

}
int lastSum(long in, int len)
{
    //exponent for finding every other number
    int x = 1;
    int sum = 0;
    //selects every other integer, starting from last
    for (int i = 0; i < len; i++)
    {
        long pwr = pow(10, x);
        long div = pow(10, x - 1);
        long current = ((in % pwr) / div);
        //result is summed together
        sum = sum + current;
        // printf("CURRENT: %li\n", current);
        // printf("SUM: %i\n", sum);
        x = x + 2;
    }
    return sum;
}